src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,605 @@
+/***************************************************************************
+                         qgsmeshtracerenderer.cpp
+                         -------------------------
+    begin                : November 2019
+    copyright            : (C) 2019 by Vincent Cloarec
+    email                : vcloarec at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmeshtracerenderer.h"
+
+
+
+QgsVector QgsMeshVectorValueInterpolatorFromVertex::value( const QgsPointXY &point )
+{
+  updateFace( point );
+
+  if ( mCacheFaceIndex == -1 || mCacheFaceIndex >= mTriangularMesh.triangles().count() )
+    return ( QgsVector( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ) );
+
+  if ( ! mActiveFaceFlagValue.active( mTriangularMesh.trianglesToNativeFaces()[mCacheFaceIndex] ) )
+    return ( QgsVector( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ) );
+
+  face = mTriangularMesh.triangles().at( mCacheFaceIndex );
+
+
+  QgsPoint p1 = mTriangularMesh.vertices().at( face[0] );
+  QgsPoint p2 = mTriangularMesh.vertices().at( face[1] );
+  QgsPoint p3 = mTriangularMesh.vertices().at( face[2] );
+
+  QgsVector v1 = QgsVector( mDatasetValues.value( face[0] ).x(),
+                            mDatasetValues.value( face[0] ).y() );
+
+  QgsVector v2 = QgsVector( mDatasetValues.value( face[1] ).x(),
+                            mDatasetValues.value( face[1] ).y() );
+
+  QgsVector v3 = QgsVector( mDatasetValues.value( face[2] ).x(),
+                            mDatasetValues.value( face[2] ).y() );
+
+
+  return QgsMeshLayerUtils::interpolateVectorFromVerticesData(
```

Comment:
note that the data can be defined also on faces (e.g. in the centre of the triangle -- one value for whole triangle area). We have a functionality to convert face-based data to vertex-based, see https://github.com/qgis/QGIS/blob/461d2edac3f092b636f114d9dbf8ec4b84e89aa8/src/core/mesh/qgsmeshlayerrenderer.cpp#L122.

Reply:
I know. I will implement a interpolator for faces too. That why there is an abstract class. There will be a interpolator class for vertex and another for faces. I will do this later, not very difficult...

Reply:
is this implemented?